### DISCUSSION TITLE
Eloquent ORM: propose to add countDistinct expression

### DISCUSSION THREAD

**Discussion by Vectorial1024** [laravel/framework#49281] [Ideas] [answered]
This seems very trivial.

Then, instead of doing this:

```php
// psalm warns about "string / implicit __toString mismatch" but it only works like this
$distinctCount = Model::query()->count(DB::raw('DISTINCT id'));
```

I can do something like:

```php
$distinctCount = Model::query()->countDistinct('id');
```

Is this unavailable because not every DB query language supports the underlying `COUNT(DISTINCT(id))` expression?

**Comment by shaedrich** [comment]
Laravel sometimes does nothing in certain cases when a database engine doesn't support a certain operation. So this hardly qualifies as a rectiffication ^^

**Comment by shaedrich** [comment]
Another interesting method would be `countGroups()` where you get back a collection with the keys being the `GROUP BY` value and the values being the result of `COUNT`.

**Comment by Vectorial1024** [comment]
From the docs https://laravel.com/docs/10.x/database Laravel currently supports 5 databases:
- MariaDB
- MySQL
- PostgreSQL
- SQLite
- SQL Server

A quick look online indicates that all the 5 databases can support `COUNT(DISTINCT)` (at least in the latest version), so in terms of database support, there is no real problem.

The question then becomes whether there are specific reasons to not do it. EG maybe this violates the Eloquent design?

**Reply by shaedrich** [reply]
The reason usually is, Taylor doesn't want to :shrug:

**Reply by Vectorial1024** [reply]
OK lol, that is one unexpected answer...

But I guess this is one valid answer.

But still, I wonder whether there are actually more people out there that also wants something like `countDistinct()`. Maybe we can get an update if there really are enough demand for this.

**Reply by shaedrich** [reply]
> OK lol, that is one unexpected answer...

Unexpected, maybe, but unfortunately, from experience 😅 

> But still, I wonder whether there are actually more people out there that also wants something like countDistinct(). Maybe we can get an update if there really are enough demand for this.

The suggestion will probably be to create a package for this or utilize a macro.

**Reply by Vectorial1024** [reply, marked as answer]
I revisited this some time later, and then I realized my smooth brain missed this very Eloquent expression that actually works:

```php
// no psalm warnings! very nice.
$distinctCount = Model::query()->distinct('id')->count();
```

Then, it actually boils down to "inadequate understanding of Laravel". And understandably, Taylor is uninterested to implement this.

Update: for more complicated queries e.g. `SELECT COUNT(*) FROM (SELECT GROUP BY HAVING)`, I can just do

```php
$distinctCount = Model::query()->fromSub( /* ... */ )->count();
```

It really is about the lack of understanding.

---

@shaedrich maybe something similar can be said for your `k, COUNT(*) ... GROUP BY k` case...? Maybe the correct (& clean) expression is possible in existing Eloquent format but we simply do not know how to make it...?

**Reply by ghost** [reply]
We overcome these kind of laravel limitations by using macros: createHydrated, withDistinctCount, runPaginationCountQueryAlsoForDistinct, getCountForPaginationAlsoForDistinct, paginateAlsoDistinct, getRelationExistenceCountDistinctQuery,hasDistinct.
While doing this, we also noticed that the distinct() method has some corner cases and special behaviors. I agree with Taylor that not every functionality should be built into laravel (to not increase its overhead I guess). Additional functionalities like this should be coded into packages or libs. And that is what we did.
